cfpb/consumerfinance.gov
# Define when to run functional tests in spec files

Currently we define what functional tests to run in GitHub Actions in [.github/workflows/functional-tests.yml](https://github.com/cfpb/consumerfinance.gov/blob/main/.github/workflows/functional-tests.yml#L65-L99). That helped us get functional tests up and running in GHA quickly, but now that list has become a pain to maintain. Instead, we want to define whether we want to skip tests in a certain environments in the spec files themselves using the `skip-test` Cypress plugin. These changes do that and also make us more consistent in how we skip tests in spec files (i.e. by always using `skip-test` instead of other methods).

---

## Additions

- Defined a new Cypress environment setting, `ENVIRONMENT: local-machine`, to be more explicit about where Cypress is running if it's not running in any of our other environments
- Added `onlyOn`s to specs we only want to run locally (i.e. not in GHA, staging, or dev environments)
- Added TODO comments to specs that are skipped to remind us to stop skipping them when they're ready to be run in other environments

## Removals

- Removed `spec` parameter in GHA functional test workflow YAML

## Changes

- Fixed some Javascript linting errors
- Changed `if ( Cypress.browser.isHeaded )` check to `skipOn( 'headless' )` in a few specs to keep the method we use to skip tests consistent everywhere
- Changed the skipping and failing pattern in the FIG spec to better reflect what we want to do with those work-in-progress tests: only run them if the FIG page exists

## How to test this PR

1. Make sure all the specs run locally with `yarn cypress run` (see GHE/CFGOV/platform/issues/4475 for a handy list of all our specs)
2. Make sure tests we want to skip in non-local environments, like GitHub Actions and staging, return as pending in those environments by appending `--env ENVIRONMENT=[whatever]` to the command, like `yarn cypress run --spec test/cypress/integration/data-research/mortgage-performance-trends/mortgage-performance-trends.cy.js --env ENVIRONMENT=staging`
3. Make sure the tests that are supposed to be skipped in headless mode are indeed skipped, like `yarn cypress run --spec test/cypress/integration/consumer-tools/youth-financial-education/survey-errors.cy.js`
4. Make sure the only FIG test that runs if you don't have a FIG page is the first test to skip all the other tests if the page doesn't exist, like `yarn cypress run --spec test/cypress/integration/compliance/small-business-lending/filing-instruction-guide.cy.js`
5. Try creating a placeholder FIG page at `/compliance/compliance-resources/small-business-lending/1071-filing-instruction-guide/`, rerun the FIG tests (`yarn cypress run --spec test/cypress/integration/compliance/small-business-lending/filing-instruction-guide.cy.js`), and make sure all the other tests start to run (they'll all fail, but that's expected without the exact right page structure)

## Notes and todos

### Todos

- [x] Is there a better way to skip all the FIG tests if the page doesn't exist? I couldn't for the life of me figure out how to get the HTTP status of a page outside of a running test in Cypress, hence the weird top-level test I made, `it( 'skips tests if page does not exist' )`. There has to be a better way to do this, but I couldn't figure it out.

### Notes

- I'm happy with the concept of only running the mortgage performance trends spec locally. We don't have test data to run them in GHA and they're mysteriously failing in pipelines, so locally is the only place they'll pass right now. I'm also fine with doing that with a combination of an `onlyOn` and a new default environment specified in our Cypress config (that will get overwritten in other environments when Cypress is run with an ENVIRONMENT argument). I'm not super stoked about having to name that new environment `local-machine`. Because of [the way the `skip-test` plugin tries to match the URL the page is running at with the Cypress `baseURL`](https://github.com/cypress-io/cypress-skip-test/blob/master/index.js#L109), I couldn't name the default environment just "local"—that matches against "localhost" and would thus run in GitHub Actions.

## Checklist

  - PR titles are used to generate the change log in [releases](../../releases); good ones make that easier to scan.
  - Consider prefixing, e.g., "Mega Menu: fix layout bug", or "Docs: Update Docker installation instructions".
  - [This repo’s docs](https://cfpb.github.io/consumerfinance.gov/) (edit the files in the `/docs` folder) – for basic, close-to-the-code docs on working with this repo
  - CFGOV/platform wiki on GHE – for internal CFPB developer guidance
  - CFPB/hubcap wiki on GHE – for internal CFPB design and content guidance
